cpp/src/arrow/array/concatenate.cc

```diff
@@ -602,8 +691,11 @@ class ConcatenateImpl {
 }  // namespace
 
 Result<std::shared_ptr<Array>> Concatenate(const ArrayVector& arrays, MemoryPool* pool) {
-  if (arrays.size() == 0) {
-    return Status::Invalid("Must pass at least one array");
+  switch (arrays.size()) {
+    case 0:
+      return Status::Invalid("Must pass at least one array");
+    case 1:
+      return arrays[0];
```

Comment:
https://github.com/apache/arrow/issues/37878

Reply:
I will undo this. This was done after the initial reviews on this PR. I was also optimizing concatenation of children values when they were a single array, but I will also get rid of that and add an explicit comment about the reason.

Reply:
I'm keeping the original code in `array_nested.cc` as `FlattenListArray` can totally return non-trimmed slices of arrays.

```cpp
  // Final attempt to avoid invoking Concatenate().
  if (non_null_fragments.size() == 1) {
    return non_null_fragments[0];
  } else if (non_null_fragments.size() == 0) {
    return MakeEmptyArray(value_array->type(), memory_pool);
  }

  return Concatenate(non_null_fragments, memory_pool);
```

So my change was just reverting the commit that added the fast path to `Concatenate()`.